Made the library thread-safe

ericek111 opened the pull request:
Hello,

I really like your work, but it didn't work well with multithreaded applications, so I fixed it. Arrowhead already is thread-safe, but I don't like Kotlin language, makes my head hurt.

By the way, I use this library with my CS:GO external cheat and the JNA wrapper for process_vm_writev really sucks - creates hundreds of thousands objects every second and there's a huge memory leak. I'll look into that.

ericek111 committed:
Made it thread-safe + minor changes.

ericek111 committed:
Heavy optimization

Native JNI functions suck - they create hundreds of thousands of objects if used too often. That is unacceptable in an external cheat, so I created my own process_vm_readv/process_vm_writev wrapper that handles that for us. Also, I find it more convenient to use long arrays instead of iovecs in Java.

Jonatino commented on src/main/java/com/github/jonatino/process/DataSource.java:
```diff
@@ -27,6 +27,8 @@
 
 	MemoryBuffer read(Pointer address, int size);
+	
+	MemoryBuffer read(Pointer address, int size, MemoryBuffer tobuf);
```
Since we added this I think we can change remove `MemoryBuffer read(Pointer address, int size);` and change

```
 	default MemoryBuffer read(long address, int size) {
  		return read(Cacheable.pointer(address), size, Cacheable.buffer(size));
  	}
```

otherwise we would need to override two methods which do the same thing.

Also what environment did you experience the object creation on? Was it Linux, Windows or Mac?

ericek111 replied:
I'd rather have `MemoryBuffer read(Pointer address, int size, MemoryBuffer tobuf)` to load data to other MemoryBuffers, so I can easily map and access the copied memory.

I use **Ubuntu MATE 16.04** with OpenJDK 1.8.0b121. JProfiler helped me found MASSIVE memory leaks and really inefficient use of memory. Garbage collector would pause my program every now and then cleaning millions of useless variables. I already solved it in [my repo](https://github.com/ericek111/Java-Memory-Manipulation/) by using my own wrapper for process_vm_readv/process_vm_writev.

Jonatino replied:
Do you have any screenshots of what objects were created?

ericek111 replied:
I backed the old code up for reference somewhere. Give me a moment.

ericek111 replied:
Okay, here are the screenshots: http://imgur.com/a/o9FLG

Btw, there's a new update for CS:GO - new map, new case, new player models! ^_^

Jonatino replied:
Okay I will take a look at performance for Mac and Unix based systems. Thanks!

ericek111 committed:
README

ericek111 committed:
README formatting

ericek111 committed:
README

ericek111 committed:
Optimized Java classes

Forgot to commit those... I'm retarded.

ericek111 committed:
Changed error

ericek111 committed:
Include native library

If running from a JAR, extract the native library, otherwise use native java.library.path.

Jonatino commented:
Going to close this for now as I want to avoid complication by adding JNI dependencies. I was able to fix the object allocation in Linux while keeping everything in Java.

Next update will include full multithreading suppose and the ability to use multiple cached buffers at the same time!

Thank you!
